tests/tests_pytorch/trainer/test_dataloaders.py

```diff
@@ -950,104 +950,87 @@ def test_dataloaders_load_only_once_no_sanity_check(tmpdir):
     assert tracker.mock_calls == expected_sequence
 
 
-@pytest.mark.parametrize("n", [1, 2])
-def test_dataloaders_load_every_n_epochs(tmpdir, n):
-    train_reload_epochs, val_reload_epochs = [], []
-
-    class TestModel(BoringModel):
-        def train_dataloader(self):
-            train_reload_epochs.append(self.current_epoch)
-            return super().train_dataloader()
-
-        def val_dataloader(self):
-            val_reload_epochs.append(self.current_epoch)
-            return super().val_dataloader()
-
-    model = TestModel()
-
-    trainer = Trainer(
-        default_root_dir=tmpdir,
-        limit_train_batches=0.3,
-        limit_val_batches=0.3,
-        reload_dataloaders_every_n_epochs=n,
-        max_epochs=5,
-    )
-
-    tracker = Mock()
-    model.train_dataloader = Mock(wraps=model.train_dataloader)
-    model.val_dataloader = Mock(wraps=model.val_dataloader)
-    model.test_dataloader = Mock(wraps=model.test_dataloader)
-
-    tracker.attach_mock(model.train_dataloader, "train_dataloader")
-    tracker.attach_mock(model.val_dataloader, "val_dataloader")
-    tracker.attach_mock(model.test_dataloader, "test_dataloader")
-
-    trainer.fit(model)
-    trainer.test(model)
-
-    # Verify the sequence
-    expected_sequence = [call.val_dataloader(), call.train_dataloader()]  # Sanity check first
-    if n == 1:
-        expected_sequence += [call.train_dataloader(), call.val_dataloader()] * 4
-    elif n == 2:
-        expected_sequence += [call.train_dataloader(), call.val_dataloader()] * 2
-    expected_sequence += [call.test_dataloader()]
-
-    assert tracker.mock_calls == expected_sequence
-
-    # Verify epoch of reloads
-    if n == 1:
-        assert train_reload_epochs == [0, 1, 2, 3, 4]
-        assert val_reload_epochs == [0, 1, 2, 3, 4]
-    elif n == 2:
-        assert train_reload_epochs == [0, 2, 4]
-        assert val_reload_epochs == [0, 2, 4]
-
-
 @pytest.mark.parametrize(
-    "n, train_reload_epochs_expect, val_reload_epochs_expect",
+    (
+        "num_sanity_val_steps, check_val_every_n_epoch, reload_dataloaders_every_n_epochs,"
+        " train_reload_epochs_expect,val_reload_epochs_expect,val_step_epochs_expect"
+    ),
     [
-        # Sanity check at epoch 0 creates a validation dataloader, but validation is
-        # checked (and in this case reloaded) every n epochs starting from epoch n-1
-        (3, [0, 2, 4, 6, 8], [0, 2, 5, 8]),
-        (5, [0, 2, 4, 6, 8], [0, 4, 9]),
+        # general case where sanity check reloads the dataloaders for validation on current_epoch=0
+        (0, 1, 1, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
+        (1, 1, 1, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9], [1, 2, 3, 4, 5, 6, 7, 8, 9], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
```

Comment:
Can you explain why there is one less reload when sanity check runs vs when it does not run?

Reply:
because this is specific to just validation (without sanity check), so this ensures in case of `epoch=0`, we don't trigger `val_dataloader` again for validation.